### ISSUE TITLE
Improved UX for multi-control plane upgrade

### ISSUE THREAD

**Issue by ostromart** [istio/istio#24360]
The [UX for upgrade design doc](https://docs.google.com/document/d/1ecMIxUBgp5VDFqrBuu1ofyTHCwDKU1hWSLhewNKWIi4/edit#heading=h.qlmnvhzb2ib2) describes in some detail the considerations for implementing the UX around multi-control plane upgrade. This issue is to capture the actual command steps for upgrade and serve as an epoch for subtasks for 1.7 MVP. 
Let's take an example of upgrading 1.6.1 to 1.7.1:

```bash
# Install the base version.
$ istioctl install --charts ./istio-1.6.1/manifests
# Install the target version. Note: ingress gateway should be excluded at this 
# stage, since it would upgrade in place. 
# This command also may need to:
#  - surface any compatibility problems against installed existing versions
#  - report any errors due to forward incompatibility of user configs
#  - warn if revision is not set and multiple control planes found
$ istioctl install --charts ./istio-1.7.1/manifests --revision 1.7.1

# Point the data plane in default namespace to 1.7.1
$ kubectl label default istio.io/rev=1.7.1

# Perform a rolling restart of some pods or deployments.
# In practice some labeled subset of pods would likely be restarted to make the 
# rollover follow 1/10/100% rule
$ kubectl rollout restart deployment/some-deployment

# Wait some time, observe system health....
# TBD: command that lists known Istio versions in the cluster and #proxies 
# connected to each.


# Promote the target version to be the default.
# Potentially upgrade ingress gateways in-place, or document steps. Depends on what 
# GW model is used. 
# This command is still TBD, may not be required. 
istioctl x set-default-revision 1.7.1

# Remove the old version. Warn if any pods still point to this control plane.
istioctl uninstall --revision 1.6.1
```

The `set-default-revision` command will do the following:
- Update webhook config
- Store some state about which control plane istioctl targets
- Possibly change telemetry "mesh" tag.

TBD. how to help with the incremental rollover - probably some script.

**Comment by erdun** [comment]
Similar to #23889

**Comment by esnible** [comment]
Missing is discussion about the IstioOperator and/or `--set` flags when Istio was original installed.

I wrote the linked UX doc when I believed `istioctl upgrade` should by default use the IstioOperator installed by a previous version.
I got strong pushback that the default should be to ignore previous customizations, but if a user wanted upgrade to behave like their current install they could explicitly use `--profile-from-cluster` https://github.com/istio/istio/issues/23443 .

**Comment by ostromart** [comment]
Interesting - I'd like to hear the arguments behind the pushback because my default path would be to use the in-cluster config too. I see user confusion in discuss.istio.io about --set and -f, so perhaps adding another source will make that worse. Maybe the solution is to explicitly state which config source is being used in the CLI.

**Comment by esnible** [comment]
@ostromart The arguments were why https://github.com/istio/istio/pull/23377 did not merge.

If I understand, the idea is that if a user installs with custom options and then just does `istioctl upgrade` every time a new version comes out they drift along and lack repeatable way to recreate their live system.  Remembering custom settings makes it too easy to skip DevOps and stick with plain old Ops.

I am not sure how to satisfy everyone.  One possibility is requiring the `--profile-from-cluster` flag.  One possibility is to have `istioctl install` / `upgrade` look at the installed manifest even if the hypothetical `--profile-from-cluster` is skipped.  If upgrading would change a setting, we would do a `profile diff --output=flags` between the profile about to be applied and the last installed profile, and require the user `--force` or answer Y/N.

**Comment by ostromart** [comment]
All right, --profile-from-cluster seems fine. I think the --set flag is an even bigger escape hatch for users being steered to gitops, but  I guess that one is indispensable for the documents and quick testing. 
I'm adding a couple of items to the issue description:
1. ingress gateways would be upgraded in place
2. istioctl set-default-revision may require additional check to ensure target version is compatible with CRs. Ideally these would show up during target `istioctl install` but an additional hook may be required.

**Comment by ostromart** [comment]
Updated the issue description based on meeting today with @esnible @richardwxn and @therealmitchconnors . 
TLDR: 
- agreed that canary becomes the master by definition when 100% of data plane is rolled over.
- install and uninstall may need additional checks to see if other revisions are present/proxies are connected to old control plane.
- may not need set-default-revision command.
- may need further work around rolling update section.

**Comment by esnible** [comment]
https://github.com/istio/istio/issues/23479 may be a dependency for this work.

That item refers to a new document https://docs.google.com/document/d/199feuH0_a4xBkQWQB3LSEMzGOEUZ7br7k1N-vngIOYw/edit

**Comment by esnible** [comment]
@richardwxn @ostromart Is this done?

**Comment by richardwxn** [comment]
@esnible yep, closing it.
